### PR TITLE
Fix local paths for TES Azure

### PR DESCRIPTION
**Pull request by vsmalladi**

### CODE REVIEW

File: plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy [hunk 1]
```diff
@@ -319,14 +320,16 @@ class AzPath implements Path {
 
 
     String toUriString() {
+        final accountName = AzConfig.getConfig().storage().accountName
         if( path.isAbsolute() ) {
-            return "${AzFileSystemProvider.SCHEME}:/${path.toString()}"
+            return "/${accountName}${path.toString()}"
```
Comment (bentsherman) [inline plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: Should there not still be the `az://` prefix here? Or you're saying that Azure will mount the files at `/<accountName>/...` ?

Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: Azure-Tes will mount these files

File: plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesExecutor.groovy [hunk 2]
```diff
@@ -158,6 +160,14 @@ class TesExecutor extends Executor implements ExtensionPoint {
...
         new TesTaskHandler(task, this)
     }
+
+    protected BashWrapperBuilder createBashWrapperBuilder(TaskRun task) {
+        // creates the wrapper script
+        final builder = new BashWrapperBuilder(task)
+        // job directives headers
+        builder.headerScript = "NXF_CHDIR=${Escape.path(task.workDir)}"
+        return builder
+    }
```
Comment (bentsherman) [inline plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesExecutor.groovy:170]: This should be done here instead: https://github.com/nextflow-io/nextflow/blob/048c3955236cb7ff1233323c55554c0b4e966bb0/plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy#L170-L172

TesBashBuilder is a subclass of BashWrapperBuilder which is specialized for TES, so something like this should do:
```groovy
    protected TesBashBuilder newTesBashBuilder(TaskRun task, String remoteBinDir) {
        final builder = new TesBashBuilder(task, remoteBinDir)
        builder.headerScript = "NXF_CHDIR=${Escape.path(task.workDir)}"
        return builder
    }
```

File: plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy (hunk 1 above)
Comment (pditommaso) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: This will break everything

File: plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesExecutor.groovy (hunk 2 above)
Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesExecutor.groovy:170]: Good call will fix

File: plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy (hunk 1 above)
Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: What do you mean @pditommaso ?

Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: Do you mean the Azure plugin will break is that it?

Comment (pditommaso) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: `toUriString` should return path with the scheme prefix, using just `toString` if you don't need that

Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: okay let me see if I can refactor accordingly

Comment (bentsherman) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: @pditommaso this is a temporary fix for testing TES Azure, the ultimate goal is to include the account name in the AZ path only when appropriate

however now that I understand it's really the local path, likely we should implement a custom file copy strategy for it:
- `AzFileCopyStrategy`: current strategy which uses `azcopy` to transfer files
- `TesAzFileCopyStrategy`: custom strategy for TES Azure which simply the files mounted at `/<account-name>`

Comment (bentsherman) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: The main problem is that TES Azure expects AZ paths in a different format from the Azure Batch executor

Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: so whats the recomendation?

Comment (vsmalladi) [reply plugins/nf-azure/src/main/nextflow/cloud/azure/nio/AzPath.groovy:325]: Agreed and we have to find a switch if its TES vs not

File: plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/client/ApiClient.java [hunk 3]
```diff
@@ -68,7 +68,7 @@
 
 public class ApiClient {
 
-    private String basePath = "http://localhost";
+    private String basePath = "/ga4gh/tes/v1";
```
Comment (MattMcL4475) [inline plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/client/ApiClient.java:71]: is this correct?  typically TES has a url like: `https://tes.azurewebsites.net/v1/tasks`

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/client/ApiClient.java:71]: Correct. This does get over written by the config. I changes from default of localhost to `/ga4gh/tes/v1` to share what the spec has an example url. I can change it back it has no implication on the code.

**Comment by vsmalladi** [comment]
@bentsherman mostly working. Just have to work on the last part of output file checks

**Comment by bentsherman** [comment]
Regarding the output file type, when you are declaring the output files:
```groovy
        task.getOutputFilesNames().each { fileName ->
            body.addOutputsItem(outItem(fileName))
        }
```

Instead you can do something like this:
```groovy
        task.getOutputsByType(FileOutParam).keySet().each { param ->
            final type = param.type
            final patterns = param.getFilePatterns(task.context, task.workDir)
            patterns.each { pattern -> body.addOutputsItem(outItem(pattern, type)) }
        }
```

No big deal if you can't get it to work, I can just add it in the main PR

File: plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy [hunk 4]
```diff
@@ -168,17 +176,31 @@ class TesTaskHandler extends TaskHandler {
...
+        String tes_end_point = executor.getEndpoint()
+        if ( tes_end_point.contains('azure.com') ) {
+            String azure_work = WORK_DIR.replace("/", "")
+            final result = new ArrayList(BashWrapperBuilder.BASH)
+            String bashAsString = result.join(' ')
+            exec.command = List.of('/bin/bash', '-c', "${bashAsString} ${azure_work}/${wrapperFile.getName()} &> ${azure_work}/${TaskRun.CMD_LOG}".toString())
+
+        }
+        else 
+            exec.command = List.of('/bin/bash', '-c', "${BashWrapperBuilder.BASH} ${wrapperFile.getName()} &> ${TaskRun.CMD_LOG}".toString())
```
Comment (bentsherman) [inline plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: @vsmalladi why is this change needed?

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: for our TES it was having issues with the nested list

`/bin/bash -c [/bin/bash -ue] .command.run &> .command.log`

So this change was to bring it back to old way of doing it without changing our Executor too much.

`/bin/bash -c /bin/bash -ue .command.run &> .command.log`

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: I'm talking about specifically this line:
```diff
+            String azure_work = WORK_DIR.replace("/", "")
```

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: Does the container not start in `/work` ?

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: ya on azure we don't start in the work directory

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: where do you start?

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: Our tesRunner, runs at root.

So then i do `NXF_CHDIR=/work` in `.command.run`

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: but the TES task submitted by Nextflow speciffies the work dir as `/work`:
```groovy
        exec.workdir = WORK_DIR
```

so why aren't you respecting that?

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: It has to do how our runner runs in the docker container. I do use it as 

`builder.headerScript = "NXF_CHDIR=${Escape.path(WORK_DIR)}"`

I tried using /work/.command.sh but seems to be an issue what we have access to

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy:198]: So you can do `bash work/.command.run` but not `bash /work/.command.run` ?

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy]: Let me look at this again.

Comment (vsmalladi) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy]: > So you can do `bash work/.command.run` but not `bash /work/.command.run` ?

Yep its somehow the runner outside the docker container can access the outfiles. Its super odd. But works.

Comment (bentsherman) [reply plugins/nf-ga4gh/src/main/nextflow/ga4gh/tes/executor/TesTaskHandler.groovy]: okay I did some cleanup, let's merge this into the main PR and discuss as a whole
